helpers/training/collate.py

```diff
@@ -528,11 +553,70 @@ def collate_fn(batch):
 
     attn_mask = None
     batch_time_ids = None
+    input_ids = None
+    extra_batch_inputs = {}
     if StateTracker.get_model_family() == "flux":
         debug_log("Compute and stack Flux time ids")
         prompt_embeds_all, add_text_embeds_all, batch_time_ids, attn_mask = (
             compute_prompt_embeddings(captions, text_embed_cache)
         )
+    elif StateTracker.get_model_family() == "omnigen":
+        # instruction, output_image = example['instruction'], example['input_images'], example['output_image']
+        omnigen_processed_embeddings = compute_prompt_embeddings(
+            captions, text_embed_cache
+        )
+        from OmniGen.processor import OmniGenCollator
+
+        attn_mask = [e.get("attention_mask") for e in omnigen_processed_embeddings]
+        attn_mask_len = len(attn_mask[0][0])
+        attn_mask = torch.stack(attn_mask, dim=0)
+
+        # we can use the OmniGenCollator.create_position to make positional ids
+        num_tokens_for_output_images = []
+        for img_size in [
+            [
+                latent_batch.shape[3] * 8,
+                latent_batch.shape[2] * 8,
+            ]
+            * len(latent_batch)
+        ]:
+            num_img_tokens = img_size[0] * img_size[1] // 16 // 16
+            num_text_tokens = attn_mask_len
+            total_num_tokens = num_img_tokens - num_text_tokens
+            num_tokens_for_output_images.append(total_num_tokens)
+        position_ids = OmniGenCollator.create_position(
+            attn_mask, num_tokens_for_output_images
+        )
+        # pad attn_mask to match the position_ids, eg. mask [1, 1, 1, 57] -> [1, 1, 1, 4097]
```

Comment:
@staoxiao this is the padding issue i mentioned that i think is where the model's loss blows up from the start

Reply:
I will take a look~